presto-spi/src/main/java/io/prestosql/spi/function/AggregateFunction.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.function;
+
+import io.prestosql.spi.block.SortOrder;
+import io.prestosql.spi.expression.ConnectorExpression;
+import io.prestosql.spi.type.Type;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import static java.util.Objects.requireNonNull;
+
+public class AggregateFunction
+{
+    // Ideally we would want this to actually be a reference to something like AggregationImplementation or aggregationFunction registry so connectors
+    // can get a one to one mapping from presto's supported aggregate function rather than relying on the name, args and output types.
+    // Something to check with Martin what he thinks is the best course here, for now just exposing the fields that the connector can use to derive if
+    // this is an agg function it can handle and information they need to push it down.
+    private final String projectionName; // may be call it assignmentId? the purpose for this is to uniquely identify an agg but we are just using the projected expression's output symbol as the name here
```

Comment:
> The `applyAggregation` call should return a list of `ConnectorExpression` in the same order as the provided list of aggregations.

And would group-bys go? At the end?

(BTW My preference is with key-ing things in a Map, as it tends to be easier to troubleshoot and harder to get wrong, but i know we already do ordinals in applyProject)

Reply:
GroupBys don’t need to go anywhere. The new projections are for the aggregation functions. The group by columns can’t be projected into something else.

The problem with using a map is that there’s nothing to key on. The name of the assignment on the engine side is meaningless to the API. You’d end up having to come up with a synthetic mapping, so you might as well use 1,2,3,...

Reply:
> GroupBys don’t need to go anywhere. The new projections are for the aggregation functions. The group by columns can’t be projected into something else.

How will the agg be pushed down in the case of query like:
```
SELECT c, count(*)
FROM t
GROUP BY c
```
?

Reply:
Given

```
Project(c, x)
    Aggregation (x = count(*), group by (c) )
        Scan(TH0)
           c = CH0
```

applyAggregation would be called with:

```
table = TH0
aggregates = [count(*)]
groupBy = { {CH0} }
```

which would return:

```
table = TH1
projections = [x]
assignments = {x = CH1}
```

And would result in the following plan:

```
Project(c, x)
    Scan(TH1)
        c = CH0
        x = CH1
```

where `CH1` is a new column associated with `count(*)`. I think the question you're asking is how is it that `CH0` is still there. The semantics of `applyAggregation` (which should be documented in the javadoc for the API) are that the columns in groupBy should be preserved in the newly produced "virtual table", with additional columns for the aggregates.

For completeness, if the query was:

```
SELECT count(*)
FROM t
GROUP BY c
```

The plan would get transformed from:

```
Project(x)
    Aggregation (x = count(*), group by (c) )
        Scan(TH0)
           c = CH0
```

into: 

```
Project(x)
    Scan(TH1)
        c = CH0
        x = CH1
```

and then column pruning would kick in and turn it into:

```
Scan(TH1)
    x = CH1
```

Reply:
> I think the question you're asking is how is it that `CH0` is still there. The semantics of `applyAggregation` (which should be documented in the javadoc for the API) are that the columns in groupBy should be preserved in the newly produced "virtual table", with additional columns for the aggregates.

I am not convinced this is a natural choice, and may be limiting in practise.

Imagine a connector where Table Handle TH0 is just a pointer to memory location containing the data (eg a cache)
In this example, `applyAggregation` would force the cache to be populated with the aggregate calculated, resulting in TH1.
Now, Column Handle CH0 "coming from" TH0 may be totally inadequate within TH1.

IMO, it would be not only more flexible, but also cleaner from API perspective, to expect we return completely new set of column handles.
Or, more precisely, leave the column handle reuse decision totally up to the connector. (If column handle is just a name, i will reuse, if it is more pointer-ish, i cannot).

Reply:
I think that case is a little far-fetched and wouldn't work in many other scenarios even today. For instance column and table handles are reused across splits. Abstraction-wise, they've always been meant to represent "a logical reference to a table or column", and in this case, the virtual table must have the columns that were used for grouping available. Splits are where things become a little more physical. Also, note that limit, sample and filter pushdown don't do this. I see you point, though -- let me think about it a bit more.

Reply:
I have updated the PR with Martin's suggestions. As per my understanding apart from adding the `projections` to `AggregationPushdownResult` this also means instead of returning a `TableScan` the rule will now have to return 

`Project
        TableScan
`

Specially to support the future use case of `sum(x+y)` -> being converted by the connector to `sum(x) + sum(y)` 

Even though `ProjectionApplicationResult` and `AggregationPushdownResult` are identical I have still kept them as separate classes so as to allow them to evolve independently without requiring SPI changes across the board. Let me know if you would rather have a single class.

Reply:
@findepi and I had an offline conversation and came to the conclusion that we'll need allow the connector to remap the column handles used for the group by keys to new column handles. This will be important for connectors that want to rewrite an aggregation to be served by a pre-materialized table. In that case, the table handle it returns could point directly at that table, and the column handles will need to be proper column handles for that table.

So, in summary, we'll need a mapping from old column handle to new column handle for the group by keys in the `AggregationApplicationResult` class, such as `Map<ColumnHandle, ColumnHandle>`.